### DISCUSSION TITLE
Is there a way to disabled sign ups, only to have invite access?

### DISCUSSION THREAD

**Discussion by benlittlenz** [supabase/supabase#3208] [Questions] [answered]
Hi, I am looking for a way to prevent sign up authentication through my web app, only having invite access through the Supabase admin panel.
Is there a way to invite a user, and when the user accepts the invitation through email, the user then set's their password and can then sign in with that password in future?

Thanks in advance!

**Comment by kangmingtay** [comment, marked as answer]
Hey @benlittlenz, the flow would look something like this:

1. Set the "Site URL" on the Authentication --> Settings page to point to your "Set password" form page. 
2. Invite a user via the Supabase admin panel
3. User clicks on the invite link. (get the <ACCESS_TOKEN> here)
4. User is redirected to your password form page to set his / her new password.

You will need to make use of the [`PUT /user`](https://github.com/supabase/gotrue#put-user) endpoint:
```
curl -X PUT 'https://abcdefghijlmnop.supabase.co/auth/v1/user' \                        
-H "Authorization: Bearer <ACCESS_TOKEN>" \
-H "apikey: <ANON_KEY>"
-d '{
  "password":  <NEW_PASSWORD>
}' \
```

where the `<ACCESS_TOKEN>` can be obtained from local storage (after the user clicks on the invite link) and the `<ANON_KEY>` can be obtained from your supabase dashboard at "Settings --> API".

Btw thanks for bringing this up! I realised the flow isn't as smooth as it should be and a better way to improve the developer experience would be to include another field on the Supabase admin panel to indicate a custom redirect endpoint for invites sent.

**Reply by benlittlenz** [reply]
Thanks a lot @kangmingtay - the flow worked perfectly.
I managed to update the user's password by using the supabase auth `update` method.
```
const token = session?.access_token;

supabase.auth.setAuth(token);
const { data, error } = await supabase.auth.update({ password: "password" });
```

**Reply by tsh356** [reply]
Agree we need a custom redirect for Invites. Site URL is used in many places and shouldn't be universally pointing to a password reset form. 

Is there a way to detect if an authenticated user has a password set?

**Reply by tsh356** [reply]
Actually, in the invite template you can append to the path:

`<p><a href="{{ .ConfirmationURL }}/set-password">Accept the invite</a></p>`

You'll then be redirected to {Site Path}/set-password

**Reply by molimat** [reply]
> Actually, in the invite template you can append to the path:
> 
> `<p><a href="{{ .ConfirmationURL }}/set-password">Accept the invite</a></p>`
> 
> You'll then be redirected to {Site Path}/set-password

just saying thank you for this.Pretty simple, but pretty smart.

**Reply by aditodkar** [reply]
@kangmingtay I have observed one problem. I am using https://supabase.com/docs/reference/javascript/auth-api-inviteuserbyemail API. When I click on accept invite link I navigate user to set password page but if user password reset fails or user tries to accept invite 2nd time then access token is not present in url. How to handle that case?

**Reply by araskavakli** [reply]
> Actually, in the invite template you can append to the path:
> 
> `<p><a href="{{ .ConfirmationURL }}/set-password">Accept the invite</a></p>`
> 
> You'll then be redirected to {Site Path}/set-password

What an elegant and simple solution 🥇

**Reply by alex1s1** [reply]
> Actually, in the invite template you can append to the path:
> 
> `<p><a href="{{ .ConfirmationURL }}/set-password">Accept the invite</a></p>`
> 
> You'll then be redirected to {Site Path}/set-password

Unfortunately, this does not help for Single Page Applications.

**Reply by victoruxui** [reply]
> > Actually, in the invite template you can append to the path:
> > `<p><a href="{{ .ConfirmationURL }}/set-password">Accept the invite</a></p>`
> > You'll then be redirected to {Site Path}/set-password
> 
> Unfortunately, this does not help for Single Page Applications.

What would be a good solution for Single Page Applications? I set this up to redirect my user to "/invite" to update password but the user gets redirected to my home screen (login)

**Reply by elemdos** [reply]
@alex1s1 @victoruxui I imagine you could use a query parameter instead (i.e. `{{ .ConfirmationURL }}?set-password`)

**Comment by molimat** [comment]
Other suggestion is to query the "type" property in the link the user uses. It shows "invite" when it is from invite and you can use it to lead the user to the password redefinition flow, per instance.

**Comment by elemdos** [comment]
If using `inviteUserByEmail`, you can set the redirect programatically: 
```
  const { data, error } = await supabaseAdmin.auth.admin.inviteUserByEmail(email, { redirectTo: `${url}/set-password?email=${email}`})
```

**Comment by shawnmclean** [comment]
Can invite by SMS work?

**Reply by kangmingtay** [reply]
@shawnmclean nope that's currently unsupported unfortunately

**Comment by BiserSirakov** [comment]
Hey all, it may sound silly but how do you get the user's session on the redirected page? 

I can only see that the access_token is passed as a hashed parameter.

note: I am using https://supabase.com/docs/reference/javascript/auth-admin-inviteuserbyemail

**Reply by Innders** [reply]
I'm having the same issue, did you ever find a solution?

**Reply by ramsrib** [reply]
This might be helpful:
```javascript
    const router = useRouter()
    const params = new URLSearchParams(router.asPath.split("#")[1])
    const access_token = params.get('access_token')
    const refresh_token = params.get('refresh_token')
    const supabase = createClientComponentClient<Database>()

    if (access_token && refresh_token) {
        supabase.auth.setSession({
            access_token,
            refresh_token
        }).then((res) => {
            if (res.error) {
                console.error("error msg", res.error.message)
            }
             console.log("session data", res.data)
        })
    }
```

**Reply by aaditeynair** [reply]
Did you ever solve this problem? Kinda facing the same situation rn

**Comment by tylernix** [comment]
I got this combination to work with the new @supabase/ssr library in Next.js 13.
Dont forget to update the supabase [email templates](https://supabase.com/docs/guides/auth/server-side/email-based-auth-with-pkce-flow-for-ssr#update-email-templates-with-url-for-api-endpoint).

`/util/supabase/server.ts`
```ts
import { Database } from '@/types/database'
import { createServerClient, type CookieOptions } from '@supabase/ssr'
import { cookies } from 'next/headers'

export const createClient = (cookieStore: ReturnType<typeof cookies>) => {

  return createServerClient<Database>(
    process.env.NEXT_PUBLIC_SUPABASE_URL!,
    process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY!,
    {
      db: {
        schema: 'public'
      },
      cookies: {
        get(name: string) {
          return cookieStore.get(name)?.value
        },
        set(name: string, value: string, options: CookieOptions) {
          try {
            cookieStore.set({ name, value, ...options })
          } catch (error) {
            // The `set` method was called from a Server Component.
            // This can be ignored if you have middleware refreshing
            // user sessions.
          }
        },
        remove(name: string, options: CookieOptions) {
          try {
            cookieStore.set({ name, value: '', ...options })
          } catch (error) {
            // The `delete` method was called from a Server Component.
            // This can be ignored if you have middleware refreshing
            // user sessions.
          }
        },
      },
    }
  )
}
```

`app/auth/confirm/route.ts`
```ts
import { createServerClient, type CookieOptions } from '@supabase/ssr'
import { type EmailOtpType } from '@supabase/supabase-js'
import { createClient } from '@/utils/supabase/server'
import { cookies } from 'next/headers'
import { redirect } from 'next/navigation'
import { NextResponse } from 'next/server'

export async function GET(request: Request) {
    const { searchParams } = new URL(request.url)
    const token_hash = searchParams.get('token_hash')
    const type = searchParams.get('type') as EmailOtpType | null
    const next = searchParams.get('next') ?? '/'

    if (token_hash && type) {
        const cookieStore = cookies()
        const supabase = createClient(cookieStore)

        const { error } = await supabase.auth.verifyOtp({
            type,
            token_hash,
        })
        if (!error) {
            return redirect(next)
        }
    }

    // return the user to an error page with some instructions
    return redirect('/auth/auth-code-error')
}
```

`app/set-password/page.tsx`
```ts
import Link from 'next/link'
import { cookies } from 'next/headers'
import { createClient } from '@/utils/supabase/server'
import { redirect } from 'next/navigation'
import { supportEmail } from '@/utils/constants'

export default async function SetPassword({
  searchParams,
}: {
  searchParams: { message: string }
}) {
  const cookieStore = cookies()
  const supabase = createClient(cookieStore)

  const { data: { user }, } = await supabase.auth.getUser()

  const setPassword = async (formData: FormData) => {
    'use server'

    const email = user?.email
    const password = formData.get('password') as string
    const cookieStore = cookies()
    const supabase = createClient(cookieStore)

    const { error } = await supabase.auth.updateUser({
      password,
    })

    if (error) {
      return redirect('/set-password?message=Could not update user password')
    }

    return redirect('/')
  }

  return (
    <div className="flex-1 flex flex-col w-full px-8 sm:max-w-md justify-center gap-2">
      <Link
        href="/"
        className="absolute left-8 top-4 py-2 px-4 rounded-md no-underline text-foreground bg-btn-background hover:bg-btn-background-hover flex items-center group text-sm"
      >
        <svg
          xmlns="http://www.w3.org/2000/svg"
          width="24"
          height="24"
          viewBox="0 0 24 24"
          fill="none"
          stroke="currentColor"
          strokeWidth="2"
          strokeLinecap="round"
          strokeLinejoin="round"
          className="mr-2 h-4 w-4 transition-transform group-hover:-translate-x-1"
        >
          <polyline points="15 18 9 12 15 6" />
        </svg>{' '}
        Back
      </Link>

      {user ?
        <form
          className="animate-in flex-1 flex flex-col w-full justify-center gap-2 text-foreground"
          action={setPassword}
        >
          <label className="text-md" htmlFor="email">
            Email
          </label>
          <input
            className="rounded-md px-4 py-2 bg-inherit border mb-6"
            name="email"
            value={user?.email}
            disabled
          />
          <label className="text-md" htmlFor="password">
            Password
          </label>
          <input
            className="rounded-md px-4 py-2 bg-inherit border mb-6"
            type="password"
            name="password"
            placeholder="••••••••"
            required
          />
          <button
            type="submit"
            className="block w-full rounded-md bg-indigo-600 px-3.5 py-2.5 text-center text-sm font-semibold text-white shadow-sm hover:bg-indigo-500 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-indigo-600"
          >
            Set Password
          </button>

          {searchParams?.message && (
            <p className="mt-4 p-4 bg-foreground/10 text-foreground text-center">
              {searchParams.message}
            </p>
          )}
        </form>
        :
        <div className="mx-auto max-w-2xl text-center flex flex-col space-y-4">
          <p className="mt-2 text-lg leading-8 text-gray-600">
            You must be logged in to use this application.
          </p>
          <p className="mt-2 text-lg leading-8 text-gray-600">
            If you do not have an account, contact <a href={`mailto:${supportEmail}`} className='text-indigo-600'>{supportEmail}</a>.
          </p>
        </div>
      }


    </div>
  )
}
```

**Reply by DophinL** [reply]
thx bro, save my day!

**Reply by haxmart** [reply]
Legend - thanks

**Reply by lchoward** [reply]
Thank you @tylernix for sharing this 🙏🏻 !

**Comment by amm297** [comment]
I am still getting this error I have the code as the project example that creates the `npx create-next-app@latest -e with-supabase` but I see the middleware is not receiving the cookie nor the headers regarding the user.

I am trying the invite through the supabase studio, creating a new project works fine

`/utils/supabase/middleware.ts`
```typescript
import { createServerClient, type CookieOptions } from '@supabase/ssr'
import { type NextRequest, NextResponse } from 'next/server'

export const createClient = (request: NextRequest) => {
  let response = NextResponse.next({ request: { headers: request.headers } })

  const supabase = createServerClient(
    process.env.NEXT_PUBLIC_SUPABASE_URL!,
    process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY!,
    {
      cookies: {
        get(name: string) {
          return request.cookies.get(name)?.value
        },
        set(name: string, value: string, options: CookieOptions) {
          // If the cookie is updated, update the cookies for the request and response
          request.cookies.set({ name, value, ...options })
          response = NextResponse.next({ request: { headers: request.headers } })
          response.cookies.set({ name, value, ...options })
        },
        remove(name: string, options: CookieOptions) {
          // If the cookie is removed, update the cookies for the request and response
          request.cookies.set({ name, value: '', ...options })
          response = NextResponse.next({ request: { headers: request.headers } })
          response.cookies.set({ name, value: '', ...options })
        },
      },
    }
  )

  return { supabase, response }
}
```

**Comment by colespencer1453** [comment]
Hello! This may be a silly question, but with the invite by email flow is there any way to allow the user to register through a social platform such as sign in with google rather than their email and password? If I'm understanding correclty in subsequent logins following the initial invite acceptance if their email is used they will not be able to sign in with Google or other social providers. Is this correct?

**Reply by Innders** [reply]
You can still use SSO even if you have signed up through an invite link, as long as the emails match. If your email is the same for Facebook, Google, Linkedin etc then logging in through each one will result in the same account, not a new one for each.
